text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+it generic now. However, it is likely that it could be made generic in
+future, backwards-compatibly.
+
+The last methods are `rng.choose(values)` and `rng.sample(n, values)`.
+They are unfortunately similar: `rng.choose(values)` is (semantically)
+just `sample(1, values.iter())` without the `Vec` allocation. It would
+be extremely nice to reconcile them but this does not seem feasible at
+the moment. NB. by taking a slice, `choose` is O(1): it generates a
+random number in-bounds of the slice and just indexes. Laying out the
+iterator-based options: taking an fully generic iterator would require
+traversing the full iterator while generating a random number at each
+step (like `sample`), taking an `ExactSizeIterator` would reduce this
+to a single random number in total and allow truncating the traversal,
+andlastly taking a `RandomAccessIterator` would give the same indexing
+behaviour as a slice. The RFC author is unsure about the best path
+here, but notes that `Iterable` may make it possible to generalise
+this backwards-compatibly in future.
+
+The current `Rng` trait has two pieces of functionality that will be
+removed (or at least made `unstable`):
+
+- `rng.gen_weight_bool(n)`, this is trivially `rng.gen(0..n) == 0`.
+- `rng.gen_ascii_chars()`, this seems relatively specialised, and the
+  full range can be simulated with `rng.gen_iter('0'..'{')` (NB. `{`
+  is the char after `z`, necessary due to exclusivity of the RHS)
+  which unfortunately picks up some punctuation in the middle, but
+  most use cases will be perfectly well served by just
+  `rng.gen_iter('a'..'{')`. We could also provide extra
+  implementations, such as `struct PrintableAscii; impl
+  Random<PrintableAscii> for char`, allowing
+  `rng.gen_iter(PrintableAscii)`.
+
+#### Alternatives
+
+This trait doesn't need to exist, everything could be free functions
+in the `rand` module.
+
+There could be a `fn by_ref<'a>(&'a mut self) -> ByRef<'a, Self>`
+function, like occurs for iterators and in IO, to provide a real type
+backing up `impl<'a, T: ?Sized + Rng> Rng for &'a mut T {}`.
+
+## Free functions
+
+[free functions]: #free-functions
+
+Top level convenience functions are provided:
+
+```rust
+fn thread_rng() -> ThreadRng {
+    // fetch the cached RNG from TLS, or create it if it doesn't exist
+}
+
+fn random<T: Random<D>, D>(data: D) -> T {
+    thread_rng().gen(data)
+}
+```
+
+### Alternatives
+
+There is valid concern that the choice of RNGs described below leaves
+users looking at jargony names for RNGs, which no particularly obvious
+default choice; possibly leading to poor selection. Functions like:
+
+```rust
+fn secure_rng() -> OsRng { ... }
+fn balanced_rng() -> ChaChaRng { ... }
+fn weak_rng() -> XorShiftRng { ... }
+```
+
+could be provided to give some guidance. As a counterpoint, the
+`collections` library does not provide `ordered_map() -> BTreeMap`,
+`unordered_map() -> HashMap`.
+
+## Error handling
```

Comment:
I'm... nervous about making these return `Result`, because *most* RNGs will never have any error conditions, and will naturally generate one of these types, so requiring that they return result imposes the performance sucks of returning larger types, and of `unwrap`. However, maybe it is not such a problem with inlining, hoping that the branch will be recognised and removed.

@aturon and I discussed this on IRC, and we thought the design that is effectively written here was alright: `next_*` return their type directly, and a `read`-like method returns `IoResult<()>` or `IoResult<usize>`, so that consumers can opt-in to reliable handling when they need it. RNGs that can legitimately have errors would just `unwrap` inside `next_*`.